### PR TITLE
GitHubBranches: Add New/Remove-GitHubRepositoryBranch Functions

### PR DESCRIPTION
**Pull request by X-Guardian**
#### Description

This PR adds the following functions to the `GitHubBranches` module:

- `New-GitHubRepositoryBranch`
- `Remove-GitHubRepositoryBranch`

#### Issues Fixed

None

#### References

[GitHub Repos API](https://developer.github.com/v3/repos/)

#### Checklist

- [x] You actually ran the code that you just wrote, especially if you did just "one last quick change".
- [x] Comment-based help added/updated, including examples.
- [x] [Static analysis](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#static-analysis) is reporting back clean.
- [x] New/changed code adheres to our [coding guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#coding-guidelines).
- [x] Changes to the manifest file follow the [manifest guidance](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#module-manifest).
- [x] Unit tests were added/updated and are all passing. See [testing guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#testing).
- [x] Relevant usage examples have been added/updated in [USAGE.md](https://github.com/microsoft/PowerShellForGitHub/blob/master/USAGE.md).
- [ ] ~~If desired, ensure your name is added to our [Contributors list](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#contributors)~~

### CODE REVIEW

File: GitHubBranches.ps1 [hunk 1]
```diff
@@ -107,3 +107,232 @@ function Get-GitHubRepositoryBranch
...
+        Remove-GitHubRepositoryBranch -OwnerName Microsoft -RepositoryName PowerShellForGitHub -Name Test
+
+        Removes the Test branch from the specified repository.
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParameterSetName = 'Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "", Justification = "Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSReviewUnusedParameter", "", Justification = "One or more parameters (like NoStatus) are only referenced by helper methods which get access to it from the stack via Get-Variable -Scope 1.")]
+    [Alias('Remove-GitHubBranch')]
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1:293]: Done

**Comment by X-Guardian** [comment]
Can we trigger the CI for this PR?

**Comment by X-Guardian** [comment]
The [Context When the origin branch cannot be found](https://dev.azure.com/ms/PowerShellForGitHub/_build/results?buildId=85827&view=logs&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=d975cbad-38f6-5497-edce-8b44859efb2f&l=98) tests are failing because I'm getting very different error objects back between PowerShell 5 and 7 when calling `Get-GitHubRepositoryBranch` with a branch name that doesn't exist.
Looking at `Invoke-GHRestMethod` in `GithubCore`, for this scenario using PowerShell 5, the `[System.Net.WebException]` code block at line 416 is being run, but with PowerShell 7, for the same error we are falling through to the final `else` block at line 459.

Adding some debug code, I can see that in PowerShell 5, `$_.Exception.PSTypeNames` at line 416 is:

```
System.Net.WebException
System.InvalidOperationException
System.SystemException
System.Exception
System.Object
```

and in PowerShell 7  `$_.Exception.PSTypeNames` at line 416 is:

```
Microsoft.PowerShell.Commands.HttpResponseException
System.Net.Http.HttpRequestException
System.Exception
System.Object
```

Adding ` -or $_.Exception -is [Microsoft.PowerShell.Commands.HttpResponseException])` to line 416 to make PowerShell 7 take the same code path as PowerShell 5, then causes PowerShell 7 to raise the following exception:

```
Get-HttpWebResponseContent: \\nas01\data\Users\Simon\Documents\GitHub\X-Guardian\PowerShellForGitHub\GitHubCore.ps1:429
Line |
 429 |  …    $rawContent = Get-HttpWebResponseContent -WebResponse $ex.Response
     |                                                             ~~~~~~~~~~~~
     | Cannot process argument transformation on parameter 'WebResponse'. Cannot convert the "StatusCode:
     | 404, ReasonPhrase: 'Not Found', Version: 1.1, Content: System.Net.Http.HttpConnectionResponseContent,
     | Headers: {   Date: Fri, 05 Jun 2020 22:17:24 GMT   Server: GitHub.com   Status: 404 Not Found
     | X-RateLimit-Limit: 5000   X-RateLimit-Remaining: 4981   X-RateLimit-Reset: 1591395451
     | X-OAuth-Scopes: admin:org, delete_repo, repo   X-Accepted-OAuth-Scopes: repo   X-GitHub-Media-Type:
     | github.nebula-preview; format=json, github.baptiste-preview; format=json, github.mercy-preview;
     | format=json   Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP,
     | X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes,
     | X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset   Access-Control-Allow-Origin: *
     | Strict-Transport-Security: max-age=31536000; includeSubdomains; preload   X-Frame-Options: deny
     | X-Content-Type-Options: nosniff   X-XSS-Protection: 1; mode=block   Referrer-Policy:
     | origin-when-cross-origin, strict-origin-when-cross-origin   Content-Security-Policy: default-src
     | 'none'   Vary: Accept-Encoding   Vary: Accept   Vary: X-Requested-With   X-GitHub-Request-Id:
     | E2FD:32667:1C74CFD:222AAEC:5EDAC474   Content-Type: application/json; charset=utf-8   Content-Length:
     | 88 }" value of type "System.Net.Http.HttpResponseMessage" to type "System.Net.HttpWebResponse".
```

I think `Invoke-GHRestMethod` and maybe other core functions needs some refactoring to generate consistent error exceptions for PowerShell 5 and 7.

**Comment by HowardWolosky** [comment]
I've opened #223 to track that problem.

**Comment by X-Guardian** [comment]
Can we trigger the CI for this PR?

**Comment by X-Guardian** [comment]
I've added the `-Force` switch to the `Remove-GitHubRepositoryBranch` function as detailed in issue #226.

**Comment by X-Guardian** [comment]
I've added temporary code to the `New-GitHubRepositoryBranch` function to work around the differences between the current PS5 and PS7 exception object as tracked in issue #223.

**Comment by X-Guardian** [comment]
Can we trigger the CI for this PR?

**Comment by X-Guardian** [comment]
Great to have a fast CI now! This PR is ready for final review.

**Comment by HowardWolosky** [comment]
> Great to have a fast CI now!

Thanks for the encouragement/push to do the extra work to make the additional accounts to allow for the parallel processing!

> This PR is ready for final review.

Will get to the review this weekend or early next week.  Thanks again for all your work here!

**Comment by X-Guardian** [comment]
Hi @HowardWolosky, I've added pipeline support to these new functions. Can we trigger the CI for this PR?

**Comment by X-Guardian** [comment]
Can we trigger the CI for this PR?

File: GitHubBranches.ps1 [hunk 2]
```diff
@@ -154,6 +154,347 @@ filter Get-GitHubRepositoryBranch
...
+        }
+        else
+        {
+            if ($_.Exception.Message -like '*Not Found*')
+            {
+                $throwObject = "Origin branch $OriginBranchName not found"
+            }
+        }
+
+        throw $throwObject
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1:327]: Done

File: GitHubBranches.ps1 [hunk 3]
```diff
@@ -154,6 +154,347 @@ filter Get-GitHubRepositoryBranch
...
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER Name
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1]: Done

File: GitHubBranches.ps1 [hunk 4]
```diff
@@ -154,6 +154,347 @@ filter Get-GitHubRepositoryBranch
...
+        GitHub.ProjectCard
+        GitHub.ProjectColumn
+        GitHub.Release
+        GitHub.Repository
+
+    .OUTPUTS
+        None
+
+    .EXAMPLE
+        Remove-GitHubRepositoryBranch  -BranchName develop -OwnerName Microsoft -RepositoryName PowerShellForGitHub
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1]: Done

File: GitHubBranches.ps1 [hunk 5]
```diff
@@ -154,6 +154,347 @@ filter Get-GitHubRepositoryBranch
...
+
+        Removes the 'develop' branch from the specified repository.
+
+    .EXAMPLE
+        Remove-GitHubRepositoryBranch  -BranchName develop -OwnerName Microsoft -RepositoryName PowerShellForGitHub -Force
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1]: Done

File: GitHubBranches.ps1 [hunk 6]
```diff
@@ -154,6 +154,347 @@ filter Get-GitHubRepositoryBranch
...
+        methods which get access to it from the stack via Get-Variable -Scope 1.")]
+    [Alias('Remove-GitHubBranch')]
+    [Alias('Delete-GitHubRepositoryBranch')]
+    [Alias('Delete-GitHubBranch')]
+    param(
+        [Parameter(
+            Mandatory,
+            ValueFromPipelineByPropertyName,
+            Position = 1)]
+        [string] $BranchName,
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubBranches.ps1]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 7]
```diff
@@ -100,13 +100,154 @@ try
...
+        Context 'When creating a new GitHub repository branch' {
+            BeforeAll {
+                $repoName = [Guid]::NewGuid().Guid
+                $originBranchName = 'master'
+                $newBranchName = 'develop'
+                $newGitHubRepositoryParms = @{
+                    RepositoryName = $repoName
+                    AutoInit = $true
+                }
+                $repo = New-GitHubRepository @newGitHubRepositoryParms
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:122]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 8]
```diff
@@ -100,13 +100,154 @@ try
...
+
+                $newGitHubRepositoryBranchParms = @{
+                    OwnerName = $script:ownerName
+                    RepositoryName = $repoName
+                    BranchName = $newBranchName
+                    OriginBranchName = $originBranchName
+                }
+                $branch = New-GitHubRepositoryBranch @newGitHubRepositoryBranchParms
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:131]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 9]
```diff
@@ -100,13 +100,154 @@ try
...
+                $branch.BranchName | Should -Be $newBranchName
+            }
+
+            It 'Should have created the branch' {
+                $getGitHubRepositoryBranchParms = @{
+                    OwnerName = $script:ownerName
+                    RepositoryName = $repoName
+                    BranchName = $newBranchName
+                }
+                { Get-GitHubRepositoryBranch @getGitHubRepositoryBranchParms } |
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:157]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 10]
```diff
@@ -100,13 +100,154 @@ try
...
+                It 'Should throw the correct exception' {
+                    $errorMessage = "Origin branch $missingOriginBranchName not found"
+
+                    $newGitHubRepositoryBranchParms = @{
+                        OwnerName = $script:ownerName
+                        RepositoryName = $repoName
+                        BranchName = $newBranchName
+                        OriginBranchName = $missingOriginBranchName
+                    }
+                    { New-GitHubRepositoryBranch @newGitHubRepositoryBranchParms } |
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:176]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 11]
```diff
@@ -100,13 +100,154 @@ try
...
+
+            Context 'When Get-GitHubRepositoryBranch throws an undefined HttpResponseException' {
+                It 'Should throw the correct exception' {
+                    $newGitHubRepositoryBranchParms = @{
+                        OwnerName = $script:ownerName
+                        RepositoryName = 'test'
+                        BranchName = 'test'
+                        OriginBranchName = 'test'
+                    }
+                    { New-GitHubRepositoryBranch @newGitHubRepositoryBranchParms } |
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:190]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 12]
```diff
@@ -100,13 +100,154 @@ try
...
+            }
+            $repo = New-GitHubRepository @newGitHubRepositoryParms
+
+            $newGitHubRepositoryBranchParms = @{
+                OwnerName = $script:ownerName
+                RepositoryName = $repoName
+                BranchName = $newBranchName
+                OriginBranchName = $originBranchName
+            }
+            $branch = New-GitHubRepositoryBranch @newGitHubRepositoryBranchParms
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:223]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 13]
```diff
@@ -100,13 +100,154 @@ try
...
+        }
+
+        It 'Should not throw an exception' {
+            $removeGitHubRepositoryBranchParms = @{
+                OwnerName = $script:ownerName
+                RepositoryName = $repoName
+                BranchName = $newBranchName
+                Confirm = $false
+            }
+            { Remove-GitHubRepositoryBranch @removeGitHubRepositoryBranchParms } |
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:238]: Done

File: Tests/GitHubBranches.tests.ps1 [hunk 14]
```diff
@@ -100,13 +100,154 @@ try
...
+                Should -Not -Throw
+        }
+
+        It 'Should have removed the branch' {
+            $getGitHubRepositoryBranchParms = @{
+                OwnerName = $script:ownerName
+                RepositoryName = $repoName
+                BranchName = $newBranchName
+            }
+            { Get-GitHubRepositoryBranch @getGitHubRepositoryBranchParms } |
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply Tests/GitHubBranches.tests.ps1:249]: Done

File: USAGE.md [hunk 15]
```diff
@@ -454,6 +457,22 @@ Disable-GitHubRepositorySecurityFix -OwnerName microsoft -RepositoryName PowerSh
 
 ----------
 
+### Branches
+
+#### Adding a new Branch to a Repository
+
+```powershell
+New-GitHubRepositoryBranch -Name develop -OwnerName Microsoft -RepositoryName PowerShellForGitHub
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply USAGE.md]: Done

File: USAGE.md [hunk 16]
```diff
@@ -454,6 +457,22 @@ Disable-GitHubRepositorySecurityFix -OwnerName microsoft -RepositoryName PowerSh
...
+```
+
+#### Removing a Branch from a Repository
+
+```powershell
+Remove-GitHubRepositoryBranch -Name develop -OwnerName Microsoft -RepositoryName PowerShellForGitHub
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply USAGE.md]: Done

**Comment by X-Guardian** [comment]
This PR is superseded by PR #256

**Pull request closed without merging**